rmw/include/rmw/rmw.h

```diff
@@ -838,15 +838,66 @@ rmw_subscription_get_actual_qos(
   const rmw_subscription_t * subscription,
   rmw_qos_profile_t * qos);
 
-/// Take an incoming message from a subscription.
+/// Take an incoming ROS message.
 /**
- * Take an incoming ROS message from a given subscription.
+ * Read a ROS message received by the given subscription, removing it from internal queues.
+ * This function will succeed even if no ROS message was received.
  *
- * \param[in] subscription The subscription object to take from.
- * \param[out] ros_message The ROS message data on success.
- * \param[out] taken Boolean flag indicating if a message was taken or not.
- * \param[in] allocation Preallocated buffer to use (may be NULL).
+ * \remarks The same ROS message cannot be taken twice.
+ *   Callers do not have to deal with duplicates.
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | Maybe
+ * Thread-Safe        | Yes
+ * Uses Atomics       | Maybe [1]
+ * Lock-Free          | Maybe [1]
+ *
+ * <i>[1] implementation defined, check implementation documentation.</i>
+ *
+ * \par Memory allocation
+ *   It is implementation defined whether memory will be allocated on take or not.
+ *   For instance, implementations that deserialize ROS messages received over
+ *   the wire may need to perform additional memory allocations when dealing with
+ *   unbounded (dynamically-sized) fields.
+ *   A subscription allocation, if provided, may or may not be used.
+ *   Check the implementation documentation to learn about memory allocation
+ *   guarantees when taking ROS messages with and without subscription allocations.
+ *
+ * \par Thread-safety
+ *   Subscriptions are thread-safe objects, and so are all operations on them except for
+ *   finalization.
+ *   Therefore, it is safe to take from the same subscription concurrently.
+ *   However, when taking regular ROS messages:
+ *   - Access to ROS messages is not synchronized.
```

Comment:
nit:
```suggestion
 *   - Access to the same ROS message is not synchronized.
```

I find that less confusing, but maybe it's just that I'm reading it incorrectly

Reply:
Yeah, I meant it in that sense. To say `things` and to say `the same thing` sounds equivalent to me, but I may be wrong.

Reply:
Well `things` implies a group, whereas `the same thing` implies a specific instance.

I'd say the suggested change is worth doing. Or you could say:

```suggestion
 *   - Access to the ROS message is not synchronized.
```

Because in this case there's only one message interacting with this function.

Reply:
Fair enough.

Reply:
Whoops! See https://github.com/ros2/rmw/pull/271/commits/005ba4062a7508129ba0d4bcc592b6160684e15a.